### DISCUSSION TITLE
PyHEP 2022 JuptyerLite + Pyodide

### DISCUSSION THREAD

**Discussion by henryiii** [henryiii/level-up-your-python#2] [General]
Taught level-up-your-python today at PyHEP using Pyodide + JupyterLite, and went pretty well. Better than when I taught my class last week and everything was breaking due to the latest release of JuptyerLite embedding Pyodide 0.21.0 and me being on Safari (whoops!). This time I patched out 0.21.2 manually and that was better. Things that popped up:

* `time.sleep` didn’t work. It just returns immediatly. That makes a few demos weird.
* Writing a file then trying to import it didn’t work. Not sure I’d expect it too, need to look into how to load from the Pyodide fs.
* Raising `StopIteration` via `next()` caused it to hang.
* Pretty much everything under “debugging” broke, but I think that’s mostly issues with JupyterLite. I did notice Rich doesn’t seem to print in color in the notebook for some reason.

Still, that’s 90% working so very nice. :)

**Comment by rth** [comment]
> time.sleep didn’t work

That one is not likely to get resolved soon. https://github.com/pyodide/pyodide/issues/2354

> Writing a file then trying to import it didn’t work. Not sure I’d expect it too, need to look into how to load from the Pyodide fs.

We had such issues in the past, I thought we had some fix for it, due to the JS FS not being fully POSIX. But maybe it's another issue.  If you have a reproducible example a new issue would be useful.

> Raising StopIteration via next() caused it to hang.

Opened https://github.com/pyodide/pyodide/issues/3097

> Pretty much everything under “debugging” broke, but I think that’s mostly issues with JupyterLite. I did notice Rich doesn’t seem to print in color in the notebook for some reason.

Maybe a JupterLite issue. Rich should work in the Pyodide REPL (or at least it used to).

**Reply by hoodmane** [reply]
>  Writing a file then trying to import it didn’t work. Not sure I’d expect it too, need to look into how to load from the Pyodide fs.

I think you need to call `importlib.invalidate_caches` before importing it.

**Reply by hoodmane** [reply]
We should add this to the FAQ.

**Comment by mrileysamc** [comment]
> Taught level-up-your-python today at PyHEP using Pyodide + JupyterLite, and went pretty well. Better than when I taught my class last week and everything was breaking due to the latest release of JuptyerLite embedding Pyodide 0.21.0 and me being on Safari (whoops!). This time I patched out 0.21.2 manually and that was better. Things that popped up:
> 
> * `time.sleep` didn’t work. It just returns immediatly. That makes a few demos weird.
> * Writing a file then trying to import it didn’t work. Not sure I’d expect it too, need to look into how to load from the Pyodide fs.
> * Raising `StopIteration` via `next()` caused it to hang.
> * Pretty much everything under “debugging” broke, but I think that’s mostly issues with JupyterLite. I did notice Rich doesn’t seem to print in color in the notebook for some reason.
> 
> Still, that’s 90% working so very nice. :)

This seems to work:
```python
import asyncio

print('test')
await asyncio.sleep(3)
print('test2')
```
